source/code/providers/support/mycnf-scanner.cpp

```diff
@@ -0,0 +1,244 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * ---------------------------------- END OF LICENSE -----------------------------
+*/
+/**
+    \file        mycnf-scanner.cpp
+
+    \brief       Scans mysql configuration to extract authentication information.
+
+    \date        2014-11-05 16:22:00
+
+*/
+/*----------------------------------------------------------------------------*/
+
+#include "mycnf-scanner.h"
+
+namespace MySQL
+{
+    static inline bool GetPathFromDirective(const std::wstring& line, SCXCoreLib::SCXFilePath& path)
+    {
+        std::vector<std::wstring> elements;
+        SCXCoreLib::StrTokenize(line, elements, L" ");
+        if ( elements.size() != 2 )
+        {
+            return false;
+        }
+
+        path.Set(elements[1]);
+        return true;
+    }
+
+    bool MySQL_ConfigurationScanner::LoadCnfFile(
+        const SCXCoreLib::SCXFilePath& file, 
+        MySQL_Authentication& auth,
+        std::set< std::wstring >& visited, 
+        std::map<unsigned int, std::wstring>& ports)
+    {
+        SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(L" loading: ", file.Get()));
+        std::pair<std::set< std::wstring >::iterator, bool> ins = visited.insert(file.Get());
+        if (!ins.second)
+        {
+            std::wstringstream ss;
+            ss << L"Already visited " << file.Get() << L". skipping";
+            SCX_LOGWARNING(m_log, ss.str());
+            return true;
+        }
+
+        std::vector< std::wstring > lines;
+        try
+        {
+            SCXCoreLib::SCXStream::NLFs nlfs;
+
+            m_extDepends->ReadAllLinesAsUTF8(file, lines, nlfs);
+        }
+        catch (SCXCoreLib::SCXException & e)
+        {
+            std::wstringstream ss;
+            ss << L"Caught " << e.What() << L" when trying to read " << file.Get() ;
+            SCX_LOGERROR(m_log, ss.str());
+            return false;
+        }
+
+        unsigned int portNum = 0;
+        std::wstring bindAddress;
+        for(std::vector< std::wstring >::const_iterator iter = lines.begin(); iter != lines.end(); ++iter)
+        {
+            std::wstring line( SCXCoreLib::StrTrim(*iter) );
+            SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(L" Line: ", line));
+            if (SCXCoreLib::StrIsPrefix(line, L"!includedir"))
+            {
+                SCXCoreLib::SCXFilePath directory;
+                if (GetPathFromDirective(line, directory))
+                {
+                    LoadCnfDirectory(directory, auth, visited, ports);
+                }
+                else
+                {
+                    SCX_LOGWARNING(m_log, SCXCoreLib::StrAppend(L" Ignoring Error in !includedir directive.", line));
```

Comment:
The particular example that I gave (spaces in directory) is not acceptable for includedir or include.
I verified it.

> In general, on lines that we parse, if we find something invalid, we should report it right there and then and probably stop reading. This makes it easier to find parsing code down the road for specific file sections and change parsing as necessary.

OK. I will change it in my next iteration. 
But I thought, in general, we must be more lenient towards the cnf file than MySQL. As in, we must never treat as error what MySQL sees as valid. But the converse is okay.
That is why I eschewed all exceptions.
 I just ignore them and processed the rest that I could process. If it is an error, we will not be able to connect.